// @databases/mysql-schema-introspect/lib/getSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sql = exports.connect = void 0;
const mysql_1 = require("@databases/mysql");
exports.connect = mysql_1.default;
Object.defineProperty(exports, "sql", { enumerable: true, get: function () { return mysql_1.sql; } });
const TableType_1 = require("./enums/TableType");
const getColumns_1 = require("./getColumns");
const getConstraints_1 = require("./getConstraints");
const getTables_1 = require("./getTables");
async function getSchema(connection, { schemaName } = {}) {
    const [tables, allColumns, allConstraints] = await Promise.all([
        (0, getTables_1.default)(connection, {
            schemaName,
            type: [TableType_1.default.BaseTable, TableType_1.default.View],
        }),
        (0, getColumns_1.default)(connection, { schemaName }),
        (0, getConstraints_1.default)(connection, { schemaName }),
    ]);
    return {
        tables: tables.map((t) => ({
            ...t,
            columns: allColumns.filter((c) => c.schemaName === t.schemaName && c.tableName === t.tableName),
            constraints: allConstraints.filter((c) => c.schemaName === t.schemaName && c.tableName === t.tableName),
        })),
    };
}
exports.default = getSchema;
