// @databases/mysql-schema-introspect/lib/enums/DataType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DataType;
(function (DataType) {
    DataType["bigint"] = "bigint";
    DataType["binary"] = "binary";
    DataType["bit"] = "bit";
    DataType["blob"] = "blob";
    DataType["char"] = "char";
    DataType["date"] = "date";
    DataType["datetime"] = "datetime";
    DataType["decimal"] = "decimal";
    DataType["double"] = "double";
    DataType["enum"] = "enum";
    DataType["float"] = "float";
    DataType["geomcollection"] = "geomcollection";
    DataType["geometry"] = "geometry";
    DataType["int"] = "int";
    DataType["json"] = "json";
    DataType["linestring"] = "linestring";
    DataType["longblob"] = "longblob";
    DataType["longtext"] = "longtext";
    DataType["mediumblob"] = "mediumblob";
    DataType["mediumint"] = "mediumint";
    DataType["mediumtext"] = "mediumtext";
    DataType["multilinestring"] = "multilinestring";
    DataType["multipoint"] = "multipoint";
    DataType["multipolygon"] = "multipolygon";
    DataType["point"] = "point";
    DataType["polygon"] = "polygon";
    DataType["set"] = "set";
    DataType["smallint"] = "smallint";
    DataType["text"] = "text";
    DataType["time"] = "time";
    DataType["timestamp"] = "timestamp";
    DataType["tinyblob"] = "tinyblob";
    DataType["tinyint"] = "tinyint";
    DataType["tinytext"] = "tinytext";
    DataType["varbinary"] = "varbinary";
    DataType["varchar"] = "varchar";
    DataType["year"] = "year";
})(DataType || (DataType = {}));
exports.default = DataType;
